File: Detectors/KnownAnomalyDetector.m
classdef KnownAnomalyDetector < AnomalyDetector
    methods
        function res = calculate(obj, ctx)
            fprintf('  [KnownAnomaly] 正在处理 KML 已知异常...\n');
            
            % 1. 检查配置
            if isempty(ctx.kmz_path) || ~exist(ctx.kmz_path, 'file')
                fprintf('    ⚠️ KML文件不存在或未选择，跳过。\n');
                res.mask = zeros(size(ctx.inROI));
                res.debug.raw = [];
                return;
            end
            
            if isempty(ctx.ref_tif_path)
                warning('    ❌ 缺少参考影像路径(ref_tif_path)，无法校正KML坐标。');
                res.mask = zeros(size(ctx.inROI));
                return;
            end
            
            % 2. 调用 KMZMaskGenerator (对应 untitled3.m Step 3.5)
            try
                radius = 3; % 默认扩充半径
                % 实例化生成器
                generator = KMZMaskGenerator(ctx.kmz_path, ctx.ref_tif_path, ctx.kmz_keywords, radius);
                
                % 运行生成
                raw_mask = generator.run();
                
                % 3. 尺寸对齐
                targetSize = size(ctx.inROI);
                if ~isequal(size(raw_mask), targetSize)
                    fprintf('    ⚠️ 尺寸不匹配 (%dx%d vs %dx%d)，调整中...\n', ...
                        size(raw_mask,1), size(raw_mask,2), targetSize(1), targetSize(2));
                    raw_mask = imresize(raw_mask, targetSize, 'nearest');
                end
                
                % 4. 结果封装
                mask = double(raw_mask);
                mask(~ctx.inROI) = 0; % 裁剪到 ROI
                
                res.mask = mask;
                res.debug.raw = raw_mask;
                fprintf('    ✅ KML异常提取完成，像素数: %d\n', sum(mask(:)>0));
                
            catch ME
                fprintf('    ❌ KML处理出错: %s\n', ME.message);
                res.mask = zeros(size(ctx.inROI));
            end
        end
    end
end

File: Python/Detectors/known_anomaly_detector.py
"""
Known Anomaly Detector for extracting anomalies from KML/KMZ files.
"""

import numpy as np
from typing import Dict, Any
from .anomaly_detector import AnomalyDetector
import sys
import os

# Add parent directory to path for imports
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from Utils.kmz_mask_generator import KMZMaskGenerator


class KnownAnomalyDetector(AnomalyDetector):
    """
    Detects known anomalies from KML/KMZ files.
    
    This detector rasterizes KML/KMZ geometry data to create a binary mask
    of known mineral deposit locations or anomalies.
    """
    
    def calculate(self, ctx) -> Dict[str, Any]:
        """
        Extract known anomalies from KML/KMZ file.
        
        Args:
            ctx: GeoDataContext object with kmz_path configuration
            
        Returns:
            Dictionary with 'mask' (binary array) and 'debug' (raw mask)
        """
        print("  [KnownAnomaly] Processing KML known anomalies...")
        
        # 1. Check configuration
        if not hasattr(ctx, 'kmz_path') or not ctx.kmz_path or not os.path.exists(ctx.kmz_path):
            print("    ⚠️ KML file does not exist or not selected, skipping.")
            return {
                'mask': np.zeros_like(ctx.in_roi, dtype=float),
                'debug': {'raw': None}
            }
        
        if not hasattr(ctx, 'ref_tif_path') or not ctx.ref_tif_path:
            print("    ❌ Missing reference image path (ref_tif_path), cannot rectify KML coordinates.")
            return {
                'mask': np.zeros_like(ctx.in_roi, dtype=float),
                'debug': {'raw': None}
            }
        
        # 2. Call KMZMaskGenerator
        try:
            radius = 3  # Default expansion radius for point features
            
            # Get keywords
            keywords = getattr(ctx, 'kmz_keywords', ['矿体投影', 'Object ID', 'ZK', '异常', '已知矿点'])
            
            # Instantiate generator
            generator = KMZMaskGenerator(ctx.kmz_path, ctx.ref_tif_path, keywords, radius)
            
            # Run generation
            raw_mask = generator.run()
            
            # 3. Size alignment
            target_size = ctx.in_roi.shape
            if raw_mask.shape != target_size:
                print(f"    ⚠️ Size mismatch ({raw_mask.shape} vs {target_size}), resizing...")
                from scipy.ndimage import zoom
                zoom_factors = (target_size[0] / raw_mask.shape[0], 
                              target_size[1] / raw_mask.shape[1])
                raw_mask = zoom(raw_mask, zoom_factors, order=0)  # Nearest neighbor
            
            # 4. Result packaging
            mask = raw_mask.astype(float)
            mask[~ctx.in_roi] = 0  # Clip to ROI
            
            print(f"    ✅ KML anomaly extraction complete, pixel count: {np.sum(mask > 0)}")
            
            return {
                'mask': mask,
                'debug': {'raw': raw_mask}
            }
            
        except Exception as e:
            print(f"    ❌ KML processing error: {str(e)}")
            return {
                'mask': np.zeros_like(ctx.in_roi, dtype=float),
                'debug': {'raw': None}
            }
